validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/SyncCommitteePerformanceTracker.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.blocks.blockbody.versions.altair.BeaconBlockBodyAltair;
+import tech.pegasys.teku.spec.datastructures.blocks.blockbody.versions.altair.SyncAggregate;
+import tech.pegasys.teku.spec.datastructures.operations.versions.altair.SyncCommitteeSignature;
+import tech.pegasys.teku.ssz.collections.SszBitvector;
+import tech.pegasys.teku.storage.client.CombinedChainDataClient;
+
+public class SyncCommitteePerformanceTracker {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final NavigableMap<UInt64, Map<UInt64, Set<Integer>>>
+      expectedSyncCommitteeParticipantsByPeriodEndEpoch = new ConcurrentSkipListMap<>();
+
+  // Slot to set of indices of validators that produced a signature for that slot
+  private final NavigableMap<UInt64, Set<UInt64>> signatureProducersBySlot =
+      new ConcurrentSkipListMap<>();
+
+  private final Spec spec;
+  private final CombinedChainDataClient combinedChainDataClient;
+
+  public SyncCommitteePerformanceTracker(
+      final Spec spec, final CombinedChainDataClient combinedChainDataClient) {
+    this.spec = spec;
+    this.combinedChainDataClient = combinedChainDataClient;
+  }
+
+  public SafeFuture<SyncCommitteePerformance> calculatePerformance(final UInt64 epoch) {
+    final Map<UInt64, Set<Integer>> expectedSyncCommitteeParticipants =
+        getPeriodEndEpoch(epoch)
+            .map(expectedSyncCommitteeParticipantsByPeriodEndEpoch::get)
+            .orElse(Collections.emptyMap());
+
+    final UInt64 lastSlotOfEpoch = spec.computeStartSlotAtEpoch(epoch.plus(1)).minus(1);
+    final Map<UInt64, Set<UInt64>> producingValidatorsBySlot =
+        signatureProducersBySlot.subMap(
+            spec.computeStartSlotAtEpoch(epoch).minusMinZero(1), true, lastSlotOfEpoch, false);
+    return calculateSyncCommitteePerformance(
+            epoch, expectedSyncCommitteeParticipants, producingValidatorsBySlot)
+        .thenPeek(__ -> clearReportedData(epoch, lastSlotOfEpoch));
+  }
+
+  private void clearReportedData(final UInt64 epoch, final UInt64 lastSlotOfEpoch) {
+    // Clear data that has been reported on.
+    signatureProducersBySlot.headMap(lastSlotOfEpoch, false).clear();
+    expectedSyncCommitteeParticipantsByPeriodEndEpoch.headMap(epoch, true).clear();
+  }
+
+  private SafeFuture<SyncCommitteePerformance> calculateSyncCommitteePerformance(
+      final UInt64 epoch,
+      final Map<UInt64, Set<Integer>> assignedSubcommitteeIndicesByValidatorIndex,
+      final Map<UInt64, Set<UInt64>> producingValidatorsBySlot) {
+
+    final int numberOfExpectedSignatures =
+        assignedSubcommitteeIndicesByValidatorIndex.values().stream().mapToInt(Set::size).sum()
```

Comment:
Not quite following these calculations.  I thought each validator would produce 1 signature per slot.  But we're summing the assigned subcommittees which makes it seem like there should be 1 signature per subcommittee?

Reply:
There's some oddity here because a validator can appear in the sync committee multiple times, including being in the same subcommittee multiple times.  Rewards are paid based on a signature being aggregated into the `SyncAggregate` and the appropriate participation bit set for each time a validator appears in the sync committee.  So if you're in a subcommittee twice, you get paid twice.

The interesting bit is that if you're in the same subcommittee multiple times, you only need to publish one signature to the gossip network and it counts as fulfilling your duties for all the positions in the subcommittee. So there's a difference between the number of things you have to publish to gossip and the number of things you get paid for doing.


I've gone with reporting on what you get paid for doing, mostly because it's simpler but also because if you miss a signature for a subcommittee you are in once you get one penalty, but if you miss for a subcommittee you're in 10 times you get 10 penalties.  I think it's useful for our reporting to accurately reflect the rewards structure so you can calculate what percent of available rewards you actually earned.

Reply:
Very interesting - might be worth some kind of comment to explain that a single signature will be counted multiple times for rewards / penalties based on subcommittee membership.

Reply:
I should also note, that the validator client only ever produces one signature for each validator - the beacon node is responsible for publishing that one signature to each applicable subnet.  So from a validator client point of view you only ever create and send one signature per slot, from a beacon node point of view you publish one signature per slot per subcommittee and from a rewards point of view you include one signature per slot per entry in the sync committee

Reply:
And yep, added a class comment to try and explain it.